### PR TITLE
Create self-delete-using-alternate-data-streams.yml

### PR DESCRIPTION
**Pull request by dstepanic**
Reference: #894
Test data: https://github.com/mandiant/capa-testfiles/pull/231

Adding capa rule for self-deletion technique. I tried to capture the main behaviors in the rule across two implementations. I am open for suggestions or improvements to the rule. Thanks in advance!

### CODE REVIEW

**Review by mr-tz — COMMENTED** [review]
nice, this may be a bit detailed and specific to the example compilation. can we relax the counts and instructions or do you think those are already focused on the core?

**Comment by dstepanic** [comment]
> nice, this may be a bit detailed and specific to the example compilation. can we relax the counts and instructions or do you think those are already focused on the core?

It's a good question, I would lean towards those as being part of the core. FWIW, the rule also triggers on the x86/x64 PoC [compiled versions](https://github.com/LloydLabs/delete-self-poc/releases/tag/v1.1). I am happy to adjust it though, I was mainly trying to make sure it wouldn't generate FP's as well.

**Comment by mr-tz** [comment]
Gotcha, it's definitely better to avoid FPs.

File: anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          - and:
+            - mnemonic: lea
+            - offset: 0x1 = BufferSize
+        - and:
+          - count(api(kernel32.CreateFile)): 2
+          - number: 0x10000 = DELETE
+        - and:
+          - instruction:
+            - mnemonic: lea
+            - offset: -0x1D
```
[the remark this answers is not in the document]
Comment (dstepanic) [reply anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml:52]: Yeh, for sure. Feel free to adjust it as well.

[the remark this answers is not in the document]
Comment (mr-tz) [reply anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml:53]: I think this is too specific here so would remove this instruction.
image

[the remark this answers is not in the document]
Comment (mr-tz) [reply anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml:53]: for dynamic we could add `FileRenameInfo = 3` and `FileDispositionInfo = 4`

**Comment by mr-tz** [comment]
What do you think of this simplification?

```yml
  features:
    - and:
      - basic block:
        - and:
          - api: kernel32.SetFileInformationByHandle
          - optional:
            - number: 3 = FileRenameInfo
      - basic block:
        - and:
          - api: kernel32.SetFileInformationByHandle
          - number: 4 = FileDispositionInfo
          - number: 1 = TRUE // fDelete.DeleteFile = TRUE;
      - and:
        - count(api(kernel32.CreateFile)): 2
        - number: 0x10000 = DELETE

```

**Comment by dstepanic** [comment]
> What do you think of this simplification?
> 
> ```yaml
>   features:
>     - and:
>       - count(api(kernel32.SetFileInformationByHandle)): 2
>       - basic block:
>         - and:
>           - api: kernel32.SetFileInformationByHandle
>           - optional:
>             - number: 3 = FileRenameInfo
>       - basic block:
>         - and:
>           - api: kernel32.SetFileInformationByHandle
>           - number: 4 = FileDispositionInfo
>           - number: 1 = TRUE // fDelete.DeleteFile = TRUE;
>       - and:
>         - count(api(kernel32.CreateFile)): 2
>         - number: 0x10000 = DELETE
> ```

Yeh, that's much better and cleaner.  Thanks for the recommendation, I will adjust and send it back. Thanks!

File: anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      - basic block:
+        - and:
+          - api: kernel32.SetFileInformationByHandle
+          - optional:
+            - number: 3 = FileRenameInfo
+      - basic block:
+        - and:
+          - api: kernel32.SetFileInformationByHandle
+          - number: 4 = FileDispositionInfo
+          - number: 1 = TRUE // fDelete.DeleteFile = TRUE;
```
Comment (mr-tz) [inline anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml:31]: last thing while we're at it, let's add an or block each and add a call subscope so this works in the dynamic flavor
so, e.g.:

```yml
      - or:
        - basic block:
          - and:
            - api: kernel32.SetFileInformationByHandle
            - optional:
              - number: 3 = FileRenameInfo
        - call:
          - and:
            - api: SetFileInformationByHandle
            - number: 3 = FileRenameInfo
```

Comment (dstepanic) [reply anti-analysis/anti-forensic/self-deletion/self-delete-using-alternate-data-streams.yml]: Ok, thanks. I tried to implement this, let me know if I missed something.

**Review by mr-tz — APPROVED** [review]
fantastic, thank you very much!

**Comment by williballenthin** [comment]
thanks @dstepanic!

**Pull request merged**